src/v/archival/ntp_archiver_service.cc

```diff
@@ -962,20 +959,32 @@ ss::future<cloud_storage::upload_result> ntp_archiver::do_upload_segment(
         }
     };
 
-    co_return co_await _remote.upload_segment(
-      get_bucket_name(),
-      path,
-      candidate.content_length,
-      std::move(reset_func),
-      fib,
-      lazy_abort_source,
-      _segment_tags);
-}
-
-ss::future<> ntp_archiver::stream_reference::maybe_close_ref() {
-    if (stream_ref.has_value()) {
-        co_await stream_ref->get().close();
-    }
+    co_return co_await _remote
+      .upload_segment(
+        get_bucket_name(),
+        path,
+        candidate.content_length,
+        std::move(reset_func),
+        fib,
+        lazy_abort_source,
+        _segment_tags)
+      .handle_exception_type([](const ss::gate_closed_exception&) mutable {
+          return cloud_storage::upload_result::cancelled;
+      })
+      .handle_exception_type([](const ss::abort_requested_exception&) mutable {
+          return cloud_storage::upload_result::cancelled;
+      })
+      .handle_exception_type([this, path](const std::exception& e) mutable {
+          vlog(_rtclog.error, "failed to upload segment {}: {}", path, e);
+          return cloud_storage::upload_result::failed;
+      })
+      .finally([&stream_state] {
+          // Cleanup the stream state if upload_segment didn't use it.
+          if (stream_state.has_value()) {
+              return stream_state->close();
+          }
+          return ss::now();
+      });
```

Comment:
nit: This is correct, but I think the slightly more Seastar idiomatic thing to do here would be to rely on the co-routine implementation converting exceptional futures to exceptions and use a try catch instead. This is open for debate though and I don't feel strongly about it.

Something like:
```
try {
co_return co_await _remote.upload_segment(...);
} catch (const ss::gate_closed_exception&) { ... }
  catch (const ss::abort_requested_exception&) { ...}
  catch (std::exception&) { ... }
  
if (stream_state.has_value()) {
  return stream_state->close();
}
```